### PR TITLE
course pages

### PR DESCRIPTION
**Pull request by wangqiang1208**
Changes in this PR:
1. course pages and routes
2. some actions in courses controller (new, create, edit, update)
3. specs

### CODE REVIEW

File: app/views/courses/edit.html.slim [hunk 1]
```diff
@@ -0,0 +1,14 @@
+= simple_form_for @course do |f|
```
Comment (lowjoel) [inline app/views/courses/edit.html.slim]: These should be internationalised.

**Comment by lowjoel** [comment]
Can rebase please? Also please merge with #31, the tests should be merged together with the whole PR...

File: app/controllers/courses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
+class CoursesController < ApplicationController
```
Comment (lowjoel) [inline app/controllers/courses_controller.rb]: I think this should be in a Courses `module`, then in routes use `scope module: 'courses'`. That way we place all the Course modules in the Courses `module`, as submodules. So we have:
```ruby
module Courses; end
class Courses::CourseControllers < ApplicationController; end
class Courses::CoursesController < CourseControllers; end
module Courses::Assessments; end
class Courses::AssessmentControllers < CourseControllers; end
class Courses::Assessments::MissionsController < AssessmentControllers; end
...
...
module Courses::Forums; end
class Courses::ForumControllers < ApplicationController
class Courses::Forums::ForumsController < ForumControllers; end
...
```
This allows us to:
 - group controllers together by feature
 - implement breadcrumbs (as per bootstrap-sass-extras)

Hopefully the class hierarchy is then clearer.

**Comment by lowjoel** [comment]
I suggest that you pull the model and controller tests from #31 and #31 will just be the view/request specs.

File: app/controllers/courses_controller.rb (hunk 2 above)
Comment (wangqiang1208) [inline app/controllers/courses_controller.rb]: Okay, Looks good.

And what are supposed to be declared inside <code>module Courses; end</code> ? And for de-coping, I think can define the outside interfaces in it ?

Comment (lowjoel) [inline app/controllers/courses_controller.rb]: The Courses `module` will be empty, it's implemented in the other files.

What's de-coping?

**Comment by wangqiang1208** [comment]
Sure, I put them separately just want to reduce the size of PR

File: app/controllers/courses_controller.rb (hunk 2 above)
Comment (wangqiang1208) [inline app/controllers/courses_controller.rb]: sorry, de-couple

**Comment by lowjoel** [comment]
I think tests+implementation counts as one feature... The PR isn't complete otherwise.

File: app/controllers/courses_controller.rb (hunk 2 above)
Comment (lowjoel) [inline app/controllers/courses_controller.rb]: Working on it.

Need to merge the course pages first, otherwise I can’t implementmy first dummy module either…

Comment (wangqiang1208) [inline app/controllers/courses_controller.rb]: Ah OK.
Btw, I also need to put these 3 guys in a folder called courses(under controllers folder), right ? 
> module Courses; end
> class Courses::Controllers < ApplicationController; end
> class Courses::CoursesController < Controllers; end

Comment (lowjoel) [inline app/controllers/courses_controller.rb]: Yep.

File: spec/support/acts_as_tenant.rb [hunk 3]
```diff
@@ -9,6 +9,7 @@ def with_tenant(tenant, &proc)
     context "with tenant #{tenant.inspect}" do |*params|
       before(:each) do
         ActsAsTenant.current_tenant = send(tenant)
+        Capybara.app_host = "http://" + send(tenant).host
```
Comment (lowjoel) [inline spec/support/acts_as_tenant.rb]: You might want to split the capybara config as a separate pr

Comment (wangqiang1208) [inline spec/support/acts_as_tenant.rb]: Hmm OK, I'll do it in a new PR.

**Comment by lowjoel** [comment]
Can you add a failing course update to test the untested line? https://coveralls.io/files/411946439

File: app/models/course.rb [hunk 4]
```diff
@@ -6,4 +6,13 @@ class Course < ActiveRecord::Base
   belongs_to :creator, class_name: User.name
   has_many :course_users, inverse_of: :course, dependent: :destroy
   has_many :users, through: :course_users
+
+  before_validation :set_default_values, on: :create
```
Comment (lowjoel) [inline app/models/course.rb]: I think this should be an `after_initialize` callback, without an `on` clause. Because when I have a `new Course`, I want the defaults already set. I do not want to wait until `save` is called for my defaults to appear, right?

File: app/views/courses/courses/edit.html.slim [hunk 5]
```diff
@@ -0,0 +1,17 @@
+h2= t('.header')
+
+= simple_form_for @course do |f|
+  = f.input :title, label: t('courses.courses.form.title_label'),
+      placeholder: t('courses.courses.form.title_placeholder')
+
+  = f.input :description, label: t('courses.courses.form.description_label'),
+      placeholder: t('courses.courses.form.description_placeholder'),
+      input_html: { rows: '5' }
```
Comment (lowjoel) [inline app/views/courses/courses/edit.html.slim]: do we need to specify? We should put this as styles, not in the HTML.

File: spec/features/course_pages_spec.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+  describe 'course editing' do
+    let(:instance) { create(:instance) }
+
+    with_tenant(:instance) do
+      let!(:course) { create(:course) }
+      before { visit edit_course_path(course) }
+
+      context 'with valid information' do
+        let(:new_title)  { 'New Title' }
```
Comment (lowjoel) [inline spec/features/course_pages_spec.rb]: extra space between `)` and `{`

**Comment by lowjoel** [comment]
please split the factory for hostname from _ to - in a separate PR, stating why that is necessary. Make this depend on that PR... thanks!

I'll go read up on the Capybara rspec thing.

File: app/controllers/courses/controllers.rb [hunk 7]
```diff
@@ -0,0 +1 @@
+class Courses::Controllers < ApplicationController; end
```
Comment (lowjoel) [inline app/controllers/courses/controllers.rb]: please add the trailing newline.

File: config/locales/en.yml [hunk 8]
```diff
@@ -20,4 +20,19 @@
...
+        notice: "Course %{title} was created."
+      update:
+        notice: "Course %{title} was updated."
+      new:
+        header: 'New Course'
+      edit:
+        header: 'Edit Course'
+      form:
+        title_label: 'Give it an awesome name'
+        title_placeholder: 'eg: Maths Universe, Geovengers'
```
Comment (lowjoel) [inline config/locales/en.yml]: strictly speaking, it's _e.g._ for _exempli gratia_

File: app/models/course.rb (hunk 4 above)
Comment (lowjoel) [inline app/models/course.rb]:
```ruby
after_initialize :defaults, if: :new_record?
```
does that work?

Comment (wangqiang1208) [inline app/models/course.rb]: OK,I searched for a while, you are right. but I think the <code>on</code> clause is still needed, because there is no reason to call this function after the record is found..

Comment (lowjoel) [inline app/models/course.rb]: use `if`.

Comment (wangqiang1208) [inline app/models/course.rb]: Yes, I didn't notice your second reply, it works.

File: spec/features/course_pages_spec.rb (hunk 6 above)
Comment (wangqiang1208) [inline spec/features/course_pages_spec.rb]: :+1:

File: spec/features/course_pages_spec.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
+require 'rails_helper'
+
+RSpec.describe 'CoursePages', type: :feature do
```
Comment (lowjoel) [inline spec/features/course_pages_spec.rb]: This should be a phrase, since it's not a class we're describing. It's kind of like a [user story](https://en.wikipedia.org/wiki/User_story).

File: app/controllers/courses/controllers.rb (hunk 7 above)
Comment (wangqiang1208) [inline app/controllers/courses/controllers.rb]: Added already, otherwise there will be an red new line symbol at the end of line.

File: spec/features/course_pages_spec.rb (hunk 9 above)
Comment (lowjoel) [inline spec/features/course_pages_spec.rb]: Not the whole 'as a <user>, I can <do something> so that <reason>', something shorter would do, maybe 'Courses management'

File: app/controllers/courses/courses_controller.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
+class Courses::CoursesController < Courses::Controllers
+  load_and_authorize_resource
+
+  def show
```
Comment (lowjoel) [inline app/controllers/courses/courses_controller.rb]: Might want to annotate these methods as `:nodoc:` since controller actions are self-explanatory.

File: app/models/course.rb [hunk 11]
```diff
@@ -5,4 +5,13 @@ class Course < ActiveRecord::Base
...
+
+  private
+
+  def set_default_values
```
Comment (lowjoel) [inline app/models/course.rb]: and document the method please...

File: spec/features/course_pages_spec.rb (hunk 9 above)
Comment (wangqiang1208) [inline spec/features/course_pages_spec.rb]: I think the title should be as short as possible, because we need to rename file name also

OK I got what you mean, 'Courses management' makes sense

File: app/models/course.rb (hunk 11 above)
Comment (wangqiang1208) [inline app/models/course.rb]: So we should document all the new methods which we added?

Comment (lowjoel) [inline app/models/course.rb]: it would be best.

File: spec/features/course_pages_spec.rb (hunk 6 above)
Comment (weiqingtoh) [inline spec/features/course_pages_spec.rb]: :+1: for sharp eyes.

File: app/controllers/courses/courses_controller.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+      render 'edit'
+    end
+  end
+
+  def destroy #:nodoc:
+  end
+
+  private
+
+  def course_params
```
Comment (wangqiang1208) [inline app/controllers/courses/courses_controller.rb]: do we need to doc this also ?  function <code>model_name_params </code> are supposed to be put in every controller.

Comment (lowjoel) [inline app/controllers/courses/courses_controller.rb]: Obvious things following convention can be marked as `:nodoc:`

Comment (wangqiang1208) [inline app/controllers/courses/courses_controller.rb]: Sure

File: app/models/course.rb [hunk 13]
```diff
@@ -6,4 +6,14 @@ class Course < ActiveRecord::Base
   belongs_to :creator, class_name: User.name
   has_many :course_users, inverse_of: :course, dependent: :destroy
   has_many :users, through: :course_users
+
+  after_initialize :set_defaults, if: :new_record?
```
Comment (lowjoel) [inline app/models/course.rb]: Put callbacks before validations and associations. I just learnt this from the Rails documentation... I'll PR the change to CONTRIBUTING after this.

Comment (wangqiang1208) [inline app/models/course.rb]: So the general sequence would be:

1. includes **
2. callbacks
3. relations

right ?

Comment (lowjoel) [inline app/models/course.rb]: 1. includes
 2. callbacks
 3. validations
 4. relations

if I read it correctly. @jsyeo?

Comment (lowjoel) [inline app/models/course.rb]: actually includes might be dangerous.

Comment (wangqiang1208) [inline app/models/course.rb]: how about attributes ? like enums

Comment (lowjoel) [inline app/models/course.rb]: after callbacks, before validations?

 1. includes
 2. callbacks
 3. attribute overrides (e.g. enums)
 4. validations
 5. relations
